### ISSUE TITLE
report invalid overload implmentation

### ISSUE THREAD

**Issue by KotlinIsland** [astral-sh/ty#1232]
### Summary

```py
from typing import overload

@overload
def f(a: int) -> int: ...

@overload
def f(a: str) -> str: ...

def f(a):
    return 1  # expect error about this implementation being invalid
```

**Comment by dhruvmanila** [comment]
Is this similar to #128 and #103 ?

**Comment by KotlinIsland** [comment]
> Is this similar to #128 and #103 ?

no, seems very different. although perhaps it would depend on return type inference

i know the impl in my egg isn't annotated, but the signatures are annotated, this is the true source of truth in an overload impl

this issue is regarding that the body of the impl conforms to the signature declarations, i.e. here the impl does not conform to signature 2

**Comment by carljm** [comment]
Does any existing type checker implement this?

It seems quite difficult to implement in full generality.

It would be easy to tell if the return type of the implementation is compatible with no overload, but the feature requested here seems like it would at least require re-checking the implementation body once per overload.

**Comment by KotlinIsland** [comment]
> it would at least require re-checking the implementation body once per overload.

something along those lines, yes
